### ISSUE TITLE
gn result is worse than bn on pascal voc

### ISSUE THREAD

**Issue by hust-kevin** [open-mmlab/mmdetection#453]
I train Faster-rcnn:R-50-FPN use bn and gn on voc 07+12, but the gn result (AP 0.5:0.95) is worse than bn, both batch size=2, 1 gpu

| Backbone                          | AP     | AP50    | AP75    | 
|:---------------------------------:|:------:|:-------:|:-------:|
| `Faster-rcnn:R-50-FPN(bn) `       | 0.501  | 0.785   | 0.542   | 
| `Faster-rcnn:R-50-FPN(gn) `       | 0.474  | 0.790   | 0.498   |

**Comment by thangvubk** [comment]
Which backbone are you using?

**Comment by hust-kevin** [comment]
@thangvubk
gn: 'open-mmlab://detectron/resnet50_gn'
bn: 'modelzoo://resnet50'

**Comment by thangvubk** [comment]
Can you share the whole config file?

**Comment by hust-kevin** [comment]
```python
# model settings
normalize=dict(type='GN', num_groups=32, frozen=False)

model = dict(
    type='FasterRCNN',
    pretrained='open-mmlab://detectron/resnet50_gn',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        style='pytorch',
        normalize=normalize), 
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        num_outs=5,
        normalize=normalize),
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_scales=[8],
        anchor_ratios=[0.5, 1.0, 2.0],
        anchor_strides=[4, 8, 16, 32, 64],
        target_means=[.0, .0, .0, .0],
        target_stds=[1.0, 1.0, 1.0, 1.0],
        use_sigmoid_cls=True),
    bbox_roi_extractor=dict(
        type='SingleRoIExtractor',
        roi_layer=dict(type='RoIAlign', out_size=7, sample_num=2),
        out_channels=256,
        featmap_strides=[4, 8, 16, 32]),
    bbox_head=dict(
        type='SharedFCBBoxHead',
        num_fcs=2,
        in_channels=256,
        fc_out_channels=1024,
        roi_feat_size=7,
        num_classes=21, 
        target_means=[0., 0., 0., 0.],
        target_stds=[0.1, 0.1, 0.2, 0.2],
        reg_class_agnostic=False,
        normalize=normalize))
# model training and testing settings
train_cfg = dict(
    rpn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.7,
            neg_iou_thr=0.3,
            min_pos_iou=0.3,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=256,
            pos_fraction=0.5,
            neg_pos_ub=-1,
            add_gt_as_proposals=False),
        allowed_border=0,
        pos_weight=-1,
        smoothl1_beta=1 / 9.0,
        debug=False),
    rcnn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.5,
            min_pos_iou=0.5,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=512,
            pos_fraction=0.25,
            neg_pos_ub=-1,
            add_gt_as_proposals=True),
        pos_weight=-1,
        debug=False))
test_cfg = dict(
    rpn=dict(
        nms_across_levels=False,
        nms_pre=2000,
        nms_post=2000,
        max_num=2000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        score_thr=0.05, nms=dict(type='nms', iou_thr=0.5), max_per_img=100)
    # soft-nms is also supported for rcnn testing
    # e.g., nms=dict(type='soft_nms', iou_thr=0.5, min_score=0.05)
)
# dataset settings 和coco不一样
dataset_type = 'VOCDataset'
data_root = '/home/xx/share/dataset/Pascal VOC/VOCdevkit/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
data = dict(
    imgs_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type='RepeatDataset',  # to avoid reloading datasets frequently
        times=3,
        dataset=dict(
            type=dataset_type,
            ann_file=[
                data_root + 'VOC2007/ImageSets/Main/trainval.txt',
                data_root + 'VOC2012/ImageSets/Main/trainval.txt'
            ],
            img_prefix=[data_root + 'VOC2007/', data_root + 'VOC2012/'],
            img_scale=(1000, 600), 
            img_norm_cfg=img_norm_cfg,
            size_divisor=32,
            flip_ratio=0.5,
            with_mask=False,
            with_crowd=True,
            with_label=True)),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'VOC2007/ImageSets/Main/test.txt',
        img_prefix=data_root + 'VOC2007/',
        img_scale=(1000, 600),
        img_norm_cfg=img_norm_cfg,
        size_divisor=32,
        flip_ratio=0,
        with_mask=False,
        with_crowd=True,
        with_label=True),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'VOC2007/ImageSets/Main/test.txt',
        img_prefix=data_root + 'VOC2007/',
        img_scale=(1000, 600),
        img_norm_cfg=img_norm_cfg,
        size_divisor=32,
        flip_ratio=0,
        with_mask=False,
        with_label=False,
        test_mode=True))
# optimizer
optimizer = dict(type='SGD', lr=0.01/8, momentum=0.9, weight_decay=0.0001) 
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy 
lr_config = dict(policy='step', step=[3])  # actual epoch = 3 * 3 = 9
checkpoint_config = dict(interval=1)
# yapf:disable
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TextLoggerHook'),
        # dict(type='TensorboardLoggerHook')
    ])
# yapf:enable
# runtime settings
total_epochs = 4  # actual epoch = 4 * 3 = 12
dist_params = dict(backend='nccl')
log_level = 'INFO'
work_dir = './work_dirs/faster_rcnn_r50_fpn_gn_1x_voc'
load_from = None
resume_from = None
workflow = [('train', 1)]

```

**Comment by thangvubk** [comment]
Detectron normalize image in a different way from pytorch model zoo. 
You may have to change ``mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)`` to ``mean=[102.9801, 115.9465, 122.7717], std=[1.0, 1.0, 1.0], to_rgb=False)``

**Comment by hust-kevin** [comment]
thanks , I will try it again, then I will tell you the result

**Comment by donglee-afar** [comment]
> thanks , I will try it again, then I will tell you the result

hi, How is your result with gn now? I tried and there is no obvious improvement.

**Comment by hust-kevin** [comment]
@thangvubk   @hellock  @yhcao6 @donglee-afar
unfortunately，the result is also bad, I think bn/gn is the normalization on hidden layer, so i think the input normalization is not so important ，so  the bad result may caused by other 

what's more,I try the gn+[ws](https://arxiv.org/pdf/1903.10520.pdf), the result is more bad. 
PS:when I train gn+ws, the input normalization   I have't changed it, I used 

```python
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
```

| Backbone                          | AP     | AP50    | AP75    |
|:---------------------------------:|:------:|:-------:|:-------:|
| `Faster-rcnn:R-50-FPN(bn) `       | 0.501  | 0.785   | 0.542   |
| `Faster-rcnn:R-50-FPN(gn) `       | 0.479  | 0.796   | 0.507   | 
| `Faster-rcnn:R-50-FPN(gn+ws) `    | 0.464  | 0.779   | 0.482   |

**Comment by thangvubk** [comment]
@hust-kevin I will try to train VOC and let you know the results.

**Comment by hust-kevin** [comment]
ok, Look forward to your reply

**Comment by thangvubk** [comment]
I test with provided script on 8GPUs. Here is the results:
VOC_BN: 79.9
VOC_GN: 79.1
They are comparable. In this case you can simply use BN instead of GN.

**Comment by hust-kevin** [comment]
@thangvubk  what's the result of ap(0.5:0.95) , In my test, AP50 is comparable,but ap (0.5:0.95) is worse

**Comment by thangvubk** [comment]
I still got comparable AP75:
VOC_BN: 51.4
VOC_GN: 51.1

**Comment by hust-kevin** [comment]
what's your config file for GN?
I just use the default config, so the same network for bn and gn

**Comment by thangvubk** [comment]
```
# model settings
normalize = dict(type='GN', num_groups=32, frozen=False)

model = dict(
    type='FasterRCNN',
    pretrained='open-mmlab://detectron/resnet50_gn',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        style='pytorch',
        normalize=normalize),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        num_outs=5,
        normalize=normalize),
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_scales=[8],
        anchor_ratios=[0.5, 1.0, 2.0],
        anchor_strides=[4, 8, 16, 32, 64],
        target_means=[.0, .0, .0, .0],
        target_stds=[1.0, 1.0, 1.0, 1.0],
        use_sigmoid_cls=True),
    bbox_roi_extractor=dict(
        type='SingleRoIExtractor',
        roi_layer=dict(type='RoIAlign', out_size=7, sample_num=2),
        out_channels=256,
        featmap_strides=[4, 8, 16, 32]),
    bbox_head=dict(
        type='ConvFCBBoxHead',
        num_shared_convs=4,
        num_shared_fcs=1,
        in_channels=256,
        conv_out_channels=256,
        fc_out_channels=1024,
        roi_feat_size=7,
        num_classes=21,
        target_means=[0., 0., 0., 0.],
        target_stds=[0.1, 0.1, 0.2, 0.2],
        reg_class_agnostic=False,
        normalize=normalize))
# model training and testing settings
train_cfg = dict(
    rpn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.7,
            neg_iou_thr=0.3,
            min_pos_iou=0.3,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=256,
            pos_fraction=0.5,
            neg_pos_ub=-1,
            add_gt_as_proposals=False),
        allowed_border=0,
        pos_weight=-1,
        smoothl1_beta=1 / 9.0,
        debug=False),
    rcnn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.5,
            min_pos_iou=0.5,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=512,
            pos_fraction=0.25,
            neg_pos_ub=-1,
            add_gt_as_proposals=True),
        pos_weight=-1,
        debug=False))
test_cfg = dict(
    rpn=dict(
        nms_across_levels=False,
        nms_pre=2000,
        nms_post=2000,
        max_num=2000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        score_thr=0.05, nms=dict(type='nms', iou_thr=0.5), max_per_img=100)
    # soft-nms is also supported for rcnn testing
    # e.g., nms=dict(type='soft_nms', iou_thr=0.5, min_score=0.05)
)
# dataset settings
dataset_type = 'VOCDataset'
data_root = 'data/VOCdevkit/'
img_norm_cfg = dict(
    mean=[102.9801, 115.9465, 122.7717], std=[1.0, 1.0, 1.0], to_rgb=False)
data = dict(
    imgs_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type='RepeatDataset',  # to avoid reloading datasets frequently
        times=3,
        dataset=dict(
            type=dataset_type,
            ann_file=[
                data_root + 'VOC2007/ImageSets/Main/trainval.txt',
                data_root + 'VOC2012/ImageSets/Main/trainval.txt'
            ],
            img_prefix=[data_root + 'VOC2007/', data_root + 'VOC2012/'],
            img_scale=(1000, 600),
            img_norm_cfg=img_norm_cfg,
            size_divisor=32,
            flip_ratio=0.5,
            with_mask=False,
            with_crowd=True,
            with_label=True)),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'VOC2007/ImageSets/Main/test.txt',
        img_prefix=data_root + 'VOC2007/',
        img_scale=(1000, 600),
        img_norm_cfg=img_norm_cfg,
        size_divisor=32,
        flip_ratio=0,
        with_mask=False,
        with_crowd=True,
        with_label=True),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'VOC2007/ImageSets/Main/test.txt',
        img_prefix=data_root + 'VOC2007/',
        img_scale=(1000, 600),
        img_norm_cfg=img_norm_cfg,
        size_divisor=32,
        flip_ratio=0,
        with_mask=False,
        with_label=False,
        test_mode=True))
# optimizer
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(policy='step', step=[3])  # actual epoch = 3 * 3 = 9
checkpoint_config = dict(interval=1)
# yapf:disable
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TextLoggerHook'),
        # dict(type='TensorboardLoggerHook')
    ])
# yapf:enable
# runtime settings
total_epochs = 4  # actual epoch = 4 * 3 = 12
dist_params = dict(backend='nccl')
log_level = 'INFO'
work_dir = './work_dirs/faster_rcnn_r50_fpn_1x_voc0712_gn'
load_from = None
resume_from = None
workflow = [('train', 1)]
```

**Comment by hust-kevin** [comment]
yes, I read the gn paper again, maybe you are right
thanks for your patient reply !!

**Comment by donglee-afar** [comment]
> yes, I read the gn paper again, maybe you are right
> thanks for your patient reply !!
我也打算实现一下gn+ws，能加个qq交流一下吗，都是关山口的校友，qq：1955662900，向大佬学习！

**Comment by breezelj** [comment]
> I still got comparable AP75:
> VOC_BN: 51.4
> VOC_GN: 51.1
> I notice that you didnt use 4Conv+1FC for boxhead. It might be the reason.
why must use the  4Conv+1FC for boxhead in GN? but using SharedFCBBoxHead in BN, thank your  @thangvubk
